File: test/c-test/foobar.erl
% Copyright (C) 2018-2020 Olivier Boudeville
%
% This file is part of the Ceylan-Seaplus tests and examples.
%
% It has been placed in the public domain.
%
% Author: Olivier Boudeville [olivier (dot) boudeville (at) esperide (dot) com]



% Erlang-side API (minimal wrapper made available to the user code) in order to
% interact with the foobar service through its driver.
%
% Relies on the service_support common facility that factors out the relevant
% cross-service, generic features.
%
% Refer to inc/foobar.h for the vanilla, C foobar API.
%
-module(foobar).


% Detailed comment for foo_status/0.
-type foo_status() :: 'low_speed' | 'moderate_speed' | 'full_speed'.


% Detailed comment for tur_status/0.
-type tur_status() :: 'tur_value' | 'non_tur_value'.


-export_type([ foo_data/0, foo_status/0, tur_status/0 ]).


% For the foo_data record/type:
-include("foobar.hrl").


% For the Seaplus support:
-include("seaplus.hrl").



% API declaration.
%
% Note that:
%  - the functions below have a spec, yet are not even defined here
%  - no service-specific start/stop functions have been introduced here


% Detailed comment for foo/1.
-spec foo( integer() ) -> integer().


% Detailed comment for bar/2.
-spec bar( float(), foo_status() ) -> foo_data().


% Detailed comment for baz/2.
-spec baz( integer(), text_utils:ustring() ) -> tur_status().


% Detailed comment for tur/0.
-spec tur() -> boolean().


% Detailed comment for frob/1.
-spec frob( tur_status() ) -> text_utils:ustring().


% Nothing more to add!

File: test/c-test/foobar/src/foobar.c
/*
 * Copyright (C) 2018-2020 Olivier Boudeville
 *
 * This file is part of the Ceylan-Seaplus tests and examples.
 *
 * It has been placed in the public domain.
 *
 * Author: Olivier Boudeville [olivier (dot) boudeville (at) esperide (dot) com]
 *
 */

/* Example of code to be gathered in a library typically called by a C-wrapper
 * available as an Erlang module.
 *
 * Having access to this source code is not necessary at all to bridge this
 * service with Erlang: this is not an element that Seaplus or the service
 * integrator control.
 *
 */


// For malloc:
#include <stdlib.h>

// For sprintf:
#include <stdio.h>

// For strlen:
#include <string.h>



// For service function and type declarations:
#include "foobar.h"



int foo( int a )
{

  if ( a != 0 )
  {
	return a + 1 ;
  }
  else
  {

	// A good old division by zero is perfect to test C-side failures:

	int b = 1 ;

	return b / a ;

  }

}



/* Ownership of returned struct is transferred to the caller, which is thus
 * expected to deallocate it at some time in the future.
 *
 */
struct foo_data * bar( double a, enum foo_status status )
{

  struct foo_data * res = malloc( sizeof( struct foo_data ) ) ;

  if ( status == moderate_speed )
  {
	res->count = (int) 2 * a ;
	res->value = -20.0 ;
  }
  else
  {
	res->count = 110 ;
	res->value = 0.0 ;
  }

  return res ;

}



// Ownership of the string transferred to the caller.
enum tur_status baz( unsigned int u, const char * m )
{

  if ( u == 10 )
	return tur_value ;

  return non_tur_value ;

}



bool tur()
{

  return true ;

}


// Ownership of the string *not* transferred to the caller.
char * frob( enum tur_status status )
{

  if ( status == tur_value )
	return "this is tur" ;

  return "this is non-tur" ;

}
